vllm/config/utils.py

```diff
@@ -18,12 +24,104 @@ def config(cls: ConfigT) -> ConfigT:
     A decorator that ensures all fields in a dataclass have default values
     and that each field has a docstring.
 
-    If a `ConfigT` is used as a CLI argument itself, the `type` keyword argument
-    provided by `get_kwargs` will be
+    If a `ConfigT` is used as a CLI argument itself, the `type`
+    keyword argument provided by `get_kwargs` will be
     `pydantic.TypeAdapter(ConfigT).validate_json(cli_arg)` which treats the
     `cli_arg` as a JSON string which gets validated by `pydantic`.
 
     Config validation is performed by the tools/validate_config.py
     script, which is invoked during the pre-commit checks.
     """
     return cls
+
+
+def normalize_value(x):
+    """Return a stable, JSON-serializable canonical form for hashing.
+
+    Order: primitives, special types (Enum, callable, torch.dtype, Path), then
+    generic containers (Mapping/Set/Sequence) with recursion.
+    """
+    # Fast path
+    if x is None or isinstance(x, (bool, int, float, str)):
+        return x
+
+    # Enums tagged by enum class to avoid collisions with primitives
+    if isinstance(x, enum.Enum):
+        enum_type = f"{x.__class__.__module__}.{x.__class__.__qualname__}"
+        return (enum_type, normalize_value(x.value))
+
+    # Callables: allow classes (types) only; reject functions/lambdas/methods
+    if isinstance(x, type):
+        module = getattr(x, "__module__", "")
+        qual = getattr(x, "__qualname__", getattr(x, "__name__", ""))
+        return ".".join([p for p in (module, qual) if p]) or repr(x)
+    if callable(x):
+        raise TypeError(
+            "normalize_value: function or callable instance unsupported")
```

Comment:
What is an example of a class that we allow? If the thing is an instance of a class, then it's actually an object with state on it.